### PR TITLE
runtime: fix useArrowFunction and noArguments

### PR DESCRIPTION
**Pull request by smorimoto**
Note: this partially introduces ES6!

### CODE REVIEW

File: runtime/jslib.js [hunk 1]
```diff
@@ -394,76 +388,55 @@ function caml_js_wrap_callback(f) {
...
-  return function () {
-    var len = arguments.length;
-    var args = new Array(len + 1);
-    args[0] = this;
-    for (var i = 0; i < len; i++) args[i + 1] = arguments[i];
-    var res = caml_callback(f, args);
+  return function (...args) {
+    const res = caml_callback(f, [this, ...args]);
     return res instanceof Function ? caml_js_wrap_callback(res) : res;
   };
```
Comment (smorimoto) [inline runtime/jslib.js:416]: In this case, the arrow function cannot be used for `this`.

File: runtime/jslib.js [hunk 2]
```diff
@@ -394,76 +388,55 @@ function caml_js_wrap_callback(f) {
...
 //Provides: caml_js_wrap_callback_strict const
 //Requires: caml_callback
 function caml_js_wrap_callback_strict(arity, f) {
-  return function () {
-    var n = arguments.length;
-    var args = new Array(arity);
-    var len = Math.min(arguments.length, arity);
-    for (var i = 0; i < len; i++) args[i] = arguments[i];
+  return (...args) => {
+    args.length = arity;
```
Comment (smorimoto) [inline runtime/jslib.js:397]: You can use this trick to change the length of an array. I will comment the benchmark later.

File: runtime/jslib.js [hunk 3]
```diff
@@ -394,76 +388,55 @@ function caml_js_wrap_callback(f) {
...
-  return function () {
-    var len = caml_js_function_arity(f) - 1;
-    var args = new Array(len + 1);
-    args[0] = this;
-    for (var i = 0; i < len; i++) args[i + 1] = arguments[i];
-    return caml_callback(f, args);
+  return function (...args) {
+    const len = caml_js_function_arity(f) - 1;
+    args.length = len;
+    return caml_callback(f, [this, ...args]);
```
Comment (smorimoto) [inline runtime/jslib.js:439]: For many arguments, using concat may have better performance than spreading.

**Comment by TyOverby** [comment]
This PR looks like it contains some formatting changes (using arrow functions) and also rewrites some of the most foundational primitives in the jsoo runtime.  Could you split these up so that that latter can be thouroughly reviewed and benchmarked?

**Comment by TyOverby** [comment]
Also, why are you interested in making the change to use arrow functions and avoid using `arguments`?

**Comment by smorimoto** [comment]
### Context Differences

- The Arrow function inherits this from its parent. This may reduce the overhead of binding this by using the Arrow function.
- The function function binds this to its own scope, so you must explicitly bind this to prevent unintended behaviour in situations where this is treated differently.

The arrow function has no binding processing, so you might see some performance improvement in this case. That said, you should not see much difference in general.

**Comment by TyOverby** [comment]
> The function function binds this to its own scope, so you must explicitly bind this to prevent unintended behaviour in situations where this is treated differently.

Does this ever come up in jsoo?

**Comment by smorimoto** [comment]
@TyOverby See `runtime/fs_fake.js`

**Comment by smorimoto** [comment]
> Also, why are you interested in using arrow functions and avoid using `arguments`?

`arguments` is not `Array`, it's `ArrayLike`, and converting it to `Array` has some cost. Also, newer JS engines have better optimisation for the rest parameter.

File: runtime/jslib.js (hunk 2 above)
Comment (smorimoto) [reply runtime/jslib.js:397]: Oops, `splice` is better than `length`: https://jsben.ch/nRrGR

Comment (smorimoto) [reply runtime/jslib.js:397]: https://jsben.ch/dsvFX

File: runtime/jslib.js (hunk 3 above)
Comment (smorimoto) [reply runtime/jslib.js]: https://jsben.ch/ZZfEd

**Comment by smorimoto** [comment]
The rest parameter and arguments show almost no difference in the benchmark results for major browsers: https://jsben.ch/BQEVR

**Comment by TyOverby** [comment]
> arguments is not Array, it's ArrayLike, and converting it to Array has some cost

But Chrome is converting it to an array under the hood too.  I benchmarked this a few years ago and found that using the spread operator had no impact on speed

**Comment by smorimoto** [comment]
The rest parameter seems to be faster for many arguments: https://jsben.ch/Krmit

**Comment by smorimoto** [comment]
> > arguments is not Array, it's ArrayLike, and converting it to Array has some cost
> 
> But Chrome is converting it to an array under the hood too. I benchmarked this a few years ago and found that using the spread operator had no impact on speed

It's true, but it has a certain advantage here because it simplifies the runtime code. If we are going to introduce const/let, ES6 is required, and if we don't see any performance regression, I think we can consider this kind of change positively.

**Comment by TyOverby** [comment]
> The rest parameter and arguments show almost no difference in the benchmark results for major browsers: https://jsben.ch/BQEVR

On my machine at least, the new code is reliably 4% slower.

> The rest parameter seems to be faster for many arguments: https://jsben.ch/Krmit

I think we should prioritize calling functions that have a small number of arguments, as small-arity functions are more common in OCaml than functions with 100+ parameters.

> It's true, but it has a certain advantage here because it simplifies the runtime code. If we are going to introduce const/let, ES6 is required, and if we don't see any performance regression, I think we can consider this kind of change positively.

I modified your benchmark so that the arrays aren't empty, and to do some fast computation with them instead of logging (logging is slow, and mucks with benchmark results), and this new benchmark shows that even for 100-parameter calls, the old method is twice as fast (on Chrome at least) than using the new syntax: https://jsben.ch/sGBgk

**Comment by smorimoto** [comment]
> I modified your benchmark so that the arrays aren't empty, and to do some fast computation with them instead of logging (logging is slow, and mucks with benchmark results), and this new benchmark shows that even for 100-parameter calls, the old method is twice as fast (on Chrome at least) than using the new syntax: https://jsben.ch/sGBgk

I'm seeing a strange result 🤔 (I'm also on the latest Chrome btw)

image

**Comment by TyOverby** [comment]
Yeah, very strange.
Screenshot from 2024-09-26 12-54-30

**Comment by smorimoto** [comment]
I ran it on the latest Chrome with Apple Silicon. What's your environment?

**Comment by smorimoto** [comment]
Here is the result from Safari on the latest iOS:
IMG_0344

**Comment by TyOverby** [comment]
> I ran it on the latest Chrome with Apple Silicon. What's your environment?

Latest Chrome x86-64

**Comment by smorimoto** [comment]
I tested it again and it's still fast in all environments. I saw that it was very slow when ads loaded, so that might be the cause?

Chrome:
image


Firefox:
image

Safari:
image

**Comment by hhugo** [comment]
> This PR looks like it contains some formatting changes (using arrow functions) and also rewrites some of the most foundational primitives in the jsoo runtime. Could you split these up so that that latter can be thouroughly reviewed and benchmarked?

@smorimoto, I think it would be good to split the two changes as suggested above so that they can be tested independently.

**Comment by smorimoto** [comment]
To fix the noArguments, you need the arrow function, and to fix the useArrowFunction, you need to remove the arguments...

**Comment by hhugo** [comment]
> To fix the noArguments, you need the arrow function, and to fix the useArrowFunction, you need to remove the arguments...

I don't see the relation, sorry. I would expect that only spread is needed to replace usage of `arguments`.

**Comment by hhugo** [comment]
> After asking for the PR to be split up, I did the benchmarking which has me pretty confident that the change should instead be abandoned, so I don’t want anyone doing the extra work of splitting the feature up if it’s not going to be merged.

Which part is slower ?

**Comment by TyOverby** [comment]
I suspect it’s the … args syntax, but I don’t know for sure.

**Comment by hhugo** [comment]
@TyOverby, Have you run more that just micro benchmarks ? Were you able to see differences with real applications ?

**Comment by TyOverby** [comment]
I haven’t run this on any real applications because we aren’t easily able to pull in jsoo branches until wasm is merged.  However, i think that microbenchmarks are sufficient to reject what is effectively a code formatting change

**Comment by hhugo** [comment]
> I haven’t run this on any real applications because we aren’t easily able to pull in jsoo branches until wasm is merged. However, i think that microbenchmarks are sufficient to reject what is effectively a code formatting change

Except this PR is not only about formatting. Also, I've been tricked many times by javascript micro benchmarks because real program would sometime show opposite result. The last one was https://github.com/ocsigen/js_of_ocaml/pull/1647 that I ended up reverting.

**Comment by TyOverby** [comment]
What is the purpose of the PR then?  Did some user of Js_of_ocaml need the runtime javascript to be more “modern” for some reason?

FWIW, I find the new style of code in  this PR to be harder to read than the original, so without a good reason to do it (like performance), i’m pretty opposed

**Comment by hhugo** [comment]
> FWIW, I find the new style of code in this PR to be harder to read than the original, so without a good reason to do it (like performance), i’m pretty opposed

This PR will not be merge in its current state (You and I both suggested to split it up). I've extracted and reworked  part of it in https://github.com/ocsigen/js_of_ocaml/pull/1740 and I personally find the code much more digest in its new form.

**Comment by smorimoto** [comment]
Benchmarking with CAMLBOY is probably useful: https://github.com/linoscope/CAMLBOY

**Comment by smorimoto** [comment]
## 5.8.2

### First Attempt

```
ROM path: ./tobu.gb
  Frames: 5000
Duration: 3.127700
     FPS: 1598.618793
```

### Second attempt

```
ROM path: ./tobu.gb
  Frames: 5000
Duration: 2.946000
     FPS: 1697.216565
```

## This PR

### First Attempt

```
ROM path: ./tobu.gb
  Frames: 5000
Duration: 2.972200
     FPS: 1682.255568
```

### Second attempt

```
ROM path: ./tobu.gb
  Frames: 5000
Duration: 2.964000
     FPS: 1686.909582
```

**Comment by smorimoto** [comment]
In fact, this PR scope is included, but neither good nor bad effects seem to happen, at least for CAMLBOY. We may need a benchmark that has a negative impact.

**Comment by hhugo** [comment]
The more I look at it the more I dislike the systematic use of arrow functions. I think we should use them only when it improves readability significantly. If one show that it improves performance, we could perform the translation automatically on the fly. I believe we already have code for that.

**Comment by hhugo** [comment]
I'm going to close this one in favor of #1740. I don't really know what do to with the systematic use of  arrow. Let reconsider later maybe

**Comment by smorimoto** [comment]
OK

**Pull request closed without merging**